torch/nn/utils/_expanded_weights/expanded_weights_impl.py

```diff
@@ -0,0 +1,78 @@
+from torch._C import _TensorBase
+import torch
+import functools
+
+from typing import Callable, Dict, cast
+
+HANDLED_FUNCTIONS: Dict[Callable, torch.autograd.Function] = {}
+
+def implements_per_sample_grads(torch_function):
+    @functools.wraps(torch_function)
+    def decorator(autograd_func):
+        HANDLED_FUNCTIONS[torch_function] = autograd_func
+        return autograd_func
+    return decorator
+
+# ExpandedWeight represents a weight (parameter) Tensor that has an expanded
+# batch dimension. Operations on the ExpandedWeight Tensor act exactly like
+# those without an expanded batch dimension but a call to .backward() populates
+# the original (unexpanded) tensor with per-sample-gradients for in the grad_sample field
+#
+# ExpandedWeight has a fallback that always fails since we cannot know what the batch
+# dimension of the input tensor is and therefore cannot know if this is a valid call
+#
+# This is a __torch_function__ object but it could have also been a Tensor Extension
+# with a dispatch key.
+class ExpandedWeight(torch.Tensor):
+    def __init__(self, orig_weight, batch_size):
+        self.batch_size = batch_size
+        self.orig_weight = orig_weight
+
+    handled_functions = HANDLED_FUNCTIONS
+
+    # needed for conv2d default kwargs
+    conv_kwarg_options = ['stride', 'padding', 'dilation', 'groups']
+    conv_kwarg_defaults = {'stride': 1, 'padding': 0, 'dilation': 1, 'groups': 1}
+
+    def __new__(cls, orig_weight, _):
+        if not isinstance(orig_weight, torch.Tensor):
+            raise RuntimeError(f"Can only make Expanded Weights of Tensors, got {type(orig_weight).__name__}")
+        if not orig_weight.requires_grad:
+            raise RuntimeError("Can only build ExpandedWeights objects of tensors that require_grad")
+        ret = torch.Tensor._make_subclass(cast(_TensorBase, cls), orig_weight, True)
+        return ret
+
+    @classmethod
+    def __torch_function__(cls, func, _, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        conv_functions = (torch.nn.functional.conv1d, torch.nn.functional.conv2d, torch.nn.functional.conv3d)
+        if func in conv_functions:
+            remaining_kwargs = 7 - len(args)
+            remaining_kwargs_options = cls.conv_kwarg_options[4 - remaining_kwargs:]
+            ordered_kwargs = tuple(kwargs.get(key, cls.conv_kwarg_defaults[key]) for key in remaining_kwargs_options)
+            return cls.handled_functions[func].apply(*(args + ordered_kwargs))
+        if func in cls.handled_functions:
+            return cls.handled_functions[func].apply(*(args + tuple(kwargs.values())))
```

Comment:
The order is determined by insertion order but because this runs through `__torch_function__` handling, that order is fixed by the generated code (i.e. users listing the kwargs in different order won't change this). The one thing with changing it to take kwargs instead is that we'll have to have some special case handling for every function like we have with conv because the arguments can either be positional args or kwargs

Reply:
Thinking about this more: we actually would be able to take kwargs for the same reason: we know that `__torch_function__` will put these kwargs in a dict, even if the user passed them positonally.

However, when I try to do this I get: `TypeError: apply() takes no keyword arguments` so we need to do it positionally